qchem/pennylane_qchem/qchem/structure.py

```diff
@@ -1167,6 +1167,105 @@ def gradient(H, x, delta=0.00529):
+    :math:`\frac{\partial^2 \hat{H}(x)}{\partial x_i \partial x_j}` of the electronic
+    Hamiltonian evaluated at the nuclear coordinates :math:`x`.
+
+    For :math:`x_i = x_j` the derivative is approximated as
+
+    .. math::
+
+        \frac{\partial^2 \hat{H}(x)}{\partial x_i^2} \approx
+        \frac{\hat{H}(x_i + \delta) - 2 \hat{H}(x) + \hat{H}(x_i - \delta)}{\delta^2},
+
+    and for :math:`x_i \neq x_j` is given by
+
+    .. math::
+
+        \frac{\partial^2 \hat{H}(x)}{\partial x_i \partial x_j} \approx
+        \frac{\hat{H}(x_i + \delta/2, x_j + \delta/2) - \hat{H}(x_i - \delta/2, x_j + \delta/2)
+        - \hat{H}(x_i + \delta/2, x_j - \delta/2) + \hat{H}(x_i - \delta/2, x_j - \delta/2)}
+        {\delta^2}.
+
+    Args:
+        H (callable): function with signature ``H(x)`` that builds the electronic
+            Hamiltonian of the molecule for a given set of nuclear coordinates ``x``
+        x (array[float]): 1D array with the coordinates in Angstroms. The size of the
+            array should be ``3*N`` where ``N`` is the number of atoms in the molecule.
+        i (int): index of the :math:`i`-th coordinate
+        j (int): index of the :math:`j`-th coordinate
+        delta (float): Step size in Angstroms used to displace the nuclear coordinates.
+            Its default value corresponds to 0.01 Bohr radii.
+
+    Returns:
+        pennylane.Hamiltonian: the second-order derivative
+        :math:`\frac{\partial^2 \hat{H}(x)}{\partial x_i \partial x_j}` of the Hamiltonian
+
+    **Example**
+
+    >>> def H(x):
+    ...     return qml.qchem.molecular_hamiltonian(['H', 'H'], x)[0]
+
+    >>> x = np.array([0., 0., 0.35, 0., 0., -0.35])
+    >>> print(second_derivative(H, x, i=0, j=3))
+    (0.5868312587039344) [I0]
+    + (0.06451526002110404) [Z0]
+    + (0.06451526002054855) [Z1]
+    + (-0.20178489650810605) [Z2]
+    + (-0.20178489650810605) [Z3]
+    + (0.019075588799274536) [Z0 Z1]
+    + (-0.0054552078951808965) [Y0 X1 X2 Y3]
+    + (0.0054552078951808965) [Y0 Y1 X2 X3]
+    + (0.0054552078951808965) [X0 X1 Y2 Y3]
+    + (-0.0054552078951808965) [X0 Y1 Y2 X3]
+    + (0.02317330168492995) [Z0 Z2]
+    + (0.017718093789749055) [Z0 Z3]
+    + (0.017718093789749055) [Z1 Z2]
+    + (0.02317330168492995) [Z1 Z3]
+    + (0.018856528057322967) [Z2 Z3]
+    """
+
+    if not callable(H):
+        error_message = (
+            "{} object is not callable. \n"
+            "'H' should be a callable function to build the electronic Hamiltonian 'H(x)'".format(
+                type(H)
+            )
+        )
+        raise TypeError(error_message)
+
+    to_bohr = 1.8897261254535
+
+    if i == j:
+        x_p = x.copy()
+        x_p[i] += delta
+
+        x_m = x.copy()
+        x_m[i] -= delta
+
+        return (H(x_p) - 2.0 * H(x) + H(x_m)) * (delta * to_bohr) ** -2
+
+    if i != j:
+        x_pp = x.copy()
+        x_pp[i] += delta * 0.5
```

Comment:
Minor point: why are we multiplying by 0.5 instead of diving by 4 in the denominator?